esp-hal/src/lcd_cam/lcd/dpi.rs

```diff
@@ -0,0 +1,827 @@
+        self.dpi
+            .lcd_cam
+            .lcd_user()
+            .read()
+            .lcd_start()
+            .bit_is_clear()
+    }
+
+    /// Stops this transfer on the spot and returns the peripheral and buffer.
+    pub fn stop(mut self) -> (Dpi<'d>, BUF) {
+        self.stop_peripherals();
+        let (dpi, view) = self.release();
+        (dpi, BUF::from_view(view))
+    }
+
+    /// Waits for the transfer to finish and returns the peripheral and buffer.
+    ///
+    /// Note: If you specified `next_frame_en` as true in [Dpi::send], you're
+    /// just waiting for a DMA error when you call this.
+    pub fn wait(mut self) -> (Result<(), DmaError>, Dpi<'d>, BUF) {
+        while !self.is_done() {
+            core::hint::spin_loop();
+        }
+
+        // Stop the DMA.
+        //
+        // If the user sends more data to the DMA than the LCD_CAM needs for a single
+        // frame, the DMA will still be running after the LCD_CAM stops.
+        self.dpi.tx_channel.stop_transfer();
+
+        // Note: There is no "done" interrupt to clear.
+
+        let (dpi, view) = self.release();
+        let result = if dpi.tx_channel.has_error() {
+            Err(DmaError::DescriptorError)
+        } else {
+            Ok(())
+        };
+
+        (result, dpi, BUF::from_view(view))
+    }
+
+    fn release(mut self) -> (Dpi<'d>, BUF::View) {
+        // SAFETY: Since forget is called on self, we know that self.dpi and
+        // self.buffer_view won't be touched again.
+        let result = unsafe {
+            let dpi = ManuallyDrop::take(&mut self.dpi);
+            let view = ManuallyDrop::take(&mut self.buffer_view);
+            (dpi, view)
+        };
+        core::mem::forget(self);
+        result
+    }
+
+    fn stop_peripherals(&mut self) {
+        // Stop the LCD_CAM peripheral.
+        self.dpi
+            .lcd_cam
+            .lcd_user()
+            .modify(|_, w| w.lcd_start().clear_bit());
+
+        // Stop the DMA
+        self.dpi.tx_channel.stop_transfer();
+    }
+}
+
+impl<'d, BUF: DmaTxBuffer> Deref for DpiTransfer<'d, BUF> {
+    type Target = BUF::View;
+
+    fn deref(&self) -> &Self::Target {
+        &self.buffer_view
+    }
+}
+
+impl<'d, BUF: DmaTxBuffer> DerefMut for DpiTransfer<'d, BUF> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.buffer_view
+    }
+}
+
+impl<'d, BUF: DmaTxBuffer> Drop for DpiTransfer<'d, BUF> {
+    fn drop(&mut self) {
+        self.stop_peripherals();
+
+        // SAFETY: This is Drop, we know that self.dpi and self.buf_view
+        // won't be touched again.
+        let view = unsafe {
+            ManuallyDrop::drop(&mut self.dpi);
+            ManuallyDrop::take(&mut self.buffer_view)
+        };
+        let _ = BUF::from_view(view);
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+/// Configuration settings for the RGB/DPI interface.
+pub struct Config {
```

Comment:
Are these configs complete, or should we mark it with `#[non_exaustive]`?

Reply:
Ehhh, the color conversion feature hasn't been implemented yet so I guess it's not complete.

`#[non_exhaustive]` is bit inconvenient to work with especially when users can still do `..Default::default()` in their code to avoid disruption.

Will update it this evening

Reply:
> #[non_exhaustive] is bit inconvenient to work with especially when users can still do ..Default::default()

I agree, it's very annoying, but it's still technically a breaking change (and we've been making enough of those :D), so let's try to avoid where we can. Long term I hope upstream rust relaxes this restriction a bit.